---
title: "Impulsivita_DADA2"
author: "Ježková"
date: "2023-11-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
##QUALITY FILTERING

```{r cars, echo=FALSE}
install.packages("dada2")
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
library(BiocManager)
BiocManager::install("dada2", force = TRUE)
library(Rcpp)
library(dada2)
library(ggplot2)

setwd("D:/SEQ/Impulsivita/Raw data/SEQ_impulsivita_orez")
list.files()
LIST<-list.files()
F_reads<-LIST[grep("-pair1.fastq.gz",LIST)]
R_reads<-LIST[grep("-pair2.fastq.gz",LIST)]
F_reads_TO<-paste0("S",F_reads)
R_reads_TO<-paste0("S",R_reads)

file.rename(from = F_reads,to=F_reads_TO)
file.rename(from = R_reads,to=R_reads_TO)

LIST<-list.files()
F_reads<-LIST[grep("-pair1.fastq.gz",LIST)]
R_reads<-LIST[grep("-pair2.fastq.gz",LIST)]

# QP.f<-plotQualityProfile(F_reads[1:10],aggregate = TRUE)+ggtitle("Forward reads")
# QP.r<-plotQualityProfile(R_reads[1:10],aggregate = TRUE)+ggtitle("Rewerse reads")

# QP.f
# QP.r

# ggsave(QP.f,filename = "D:/SEQ/Impulsivita/Forward_reads_Quality.png",width= 12, height = 10, units = "cm")
# ggsave(QP.r,filename = "D:/SEQ/Impulsivita/Revers_reads_Quality.png",width= 12, height = 10, units = "cm")

sample.names<-gsub("_trus-trimmed-pair1.fastq.gz","",F_reads)
sample.names<-gsub("-assigned-","",sample.names)
filtFs <- paste0(sample.names, "_READ1_filt.fastq.gz")
filtRs <- paste0(sample.names, "_READ2_filt.fastq.gz")


#Quality filtering
# for(x in 1:length(F_reads)) {
#   print(sample.names[x])
#   fastqPairedFilter(c(F_reads[x], R_reads[x]), c(filtFs[x], filtRs[x]),
#                     maxN=0, maxEE=2, minQ=2,truncQ=2,
#                     compress=TRUE, verbose=TRUE,
#                     minLen = c(270,190),truncLen = c(270,190))
# }
```

##DADA DENOISING

```{r pressure, echo=FALSE}
#These commands denoise quality-filtered fastq files and build abundance matrix,
#(samples in rows, ASVs in columns)

list.files("D:/SEQ/Impulsivita/Raw data/SEQ_impulsivita_orez")

#List of quality filtered fastq files
fns <- list.files()
fastqs <- fns[grepl("filt.fastq.gz", fns)]
fastqs <- sort(fastqs) 

fnFs <- fastqs[grepl("_trus[12]R_READ1_filt.fastq.gz", fastqs)] 
fnRs <- fastqs[grepl("_trus[12]R_READ2_filt.fastq.gz", fastqs)] 
sample.names <- gsub("_READ1_filt.fastq.gz","",fnFs)

#fastq dereplication
# derepFs <- derepFastq(fnFs,verbose=T)
# derepRs <- derepFastq(fnRs,verbose=T)

# names(derepFs) <- sample.names
# names(derepRs) <- sample.names

#deoising
# dadaFs <- dada(derepFs, selfConsist = TRUE,MAX_CONSIST=20,multithread = TRUE)
# dadaRs <- dada(derepRs, selfConsist = TRUE,MAX_CONSIST=20,multithread = TRUE)

#merge denoised forward and reverse ASVs
# mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE, minOverlap = 10,maxMismatch=0,justConcatenate=F)

#abundance matrix
# seqtab <- makeSequenceTable(mergers)

# unname(seqtab[1:10,1:10])
# save(seqtab,file = "D:/SEQ/Impulsivita/Výsledky/seqtab.R")

```

##CHIMERAS AND TAXONOMY
```{r kitty, echo=FALSE}
library(phyloseq)
library(ShortRead)
library(Rcpp)
library(dada2)
library(ggplot2)

load("D:/SEQ/Impulsivita/Výsledky/seqtab.R")
setwd("D:/SEQ/Impulsivita/Výsledky/")

#extraxt ASVs fasta from abundance matrix
FASTA<-DNAStringSet(colnames(seqtab))
names(FASTA)<-colnames(seqtab)
writeFasta(FASTA,"haplo.fasta")

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
sum(seqtab.nochim)/sum(seqtab)

# database availibility here: https://benjjneb.github.io/dada2/training.html

# taxa <- assignTaxonomy(seqtab.nochim, 
                       # refFasta="D:/SEQ/Impulsivita/Výsledky/silva_nr99_v138.1_train_set.fa.gz", 
                       # multithread=TRUE,minBoot = 80)
# head(unname(taxa))

#OTU TABLE
# seqtab<-otu_table(seqtab.nochim,taxa_are_rows = F)

#HAPLO
# HAPLO<-readDNAStringSet("haplo.fasta")

#TAXO
# TAXO<-tax_table(taxa)

# PHYLOSEQ<-merge_phyloseq(seqtab,TAXO,HAPLO)
# PHYLOSEQ_dupl<-PHYLOSEQ
# sample_names(PHYLOSEQ)
# save(PHYLOSEQ_dupl,file = "D:/SEQ/Impulsivita/Výsledky/PHYLOSEQ_dupl_bez META.R")
```

##METADATA
```{r ballon, echo=FALSE}
load("D:/SEQ/Impulsivita/Výsledky/PHYLOSEQ_dupl_bez META.R")
setwd("D:/SEQ/Impulsivita/")
list.files()
META<-read.delim("metadata_násilí_komplet.txt")
META$ID_individual
sample_names(PHYLOSEQ_dupl)
END<-sapply(strsplit(META$Sample,"_"),function(x) x[4],simplify = T)
NEW_NAME<-paste0("S",META$ID_individual,"_n_F_",END)
sample_names(PHYLOSEQ_dupl)%in%NEW_NAME
sample_names(PHYLOSEQ_dupl)[!sample_names(PHYLOSEQ_dupl)%in%NEW_NAME]

# spojováni metadat s PHYLOSEQ
META<-sample_data(META)
sample_data(META)
sample_names(META)<-NEW_NAME
PHYLOSEQ_dupl<-merge_phyloseq(PHYLOSEQ_dupl,META)

# save(PHYLOSEQ_dupl,file = "D:/SEQ/Impulsivita/Výsledky/PHYLOSEQ_dupl.R")

```

##DUPLICATES MERGER 
```{r jaws, echo=FALSE}
library(phyloseq)
library(vegan)
library(ggplot2)
library(ape)
library(Polychrome)
library(plyr)
library(performance)
library(tidyverse)

dupl.concensus<-function(PHYLOS,NAMES){
  
  # exclude nonduplicated samples
  IDS<-as.character(data.frame(sample_data(PHYLOS))[,NAMES])
  IDS.dupl<-IDS[duplicated(IDS)]
  
  PHYLOSEQ<-prune_samples(IDS%in%IDS.dupl, PHYLOS)
  if(length(IDS.dupl)*2<length(IDS)) {NONUPLICATED<-prune_samples(!IDS%in%IDS.dupl, PHYLOS)
  print(paste("Following names are nonduplicated",sample_names(NONUPLICATED)))}
  
  CATS<-as.character(data.frame(sample_data(PHYLOSEQ))[,NAMES])
  CATS2<-levels(factor(CATS))
  OTU_TAB<-otu_table(PHYLOSEQ)
  rownames(OTU_TAB)<-CATS
  
  # i<-5
  for (i in 1:length(CATS2))
  {
    # print(CATS2[i])
    FILTER.act<-colSums(OTU_TAB[rownames(OTU_TAB)==CATS2[i],]>0)>1
    OTU_TAB[rownames(OTU_TAB)==CATS2[i],]
    OTU_TAB[rownames(OTU_TAB)==CATS2[i],]<-t(apply(OTU_TAB[rownames(OTU_TAB)==CATS2[i],],1,function(x) x*FILTER.act))
  }
  
  rownames(OTU_TAB)<-sample_names(PHYLOSEQ)
  otu_table(PHYLOSEQ)<-OTU_TAB
  PHYLOSEQ.clean<-prune_taxa(taxa_sums(PHYLOSEQ)>0,PHYLOSEQ)
  
  PHYLOSEQ.clean
}

#This script merge technical duplicates (specifired in "NAMES" argument)
merge.duplicates<-function(PHYLOSEQ,NAMES){
  CATS<-as.character(data.frame(sample_data(PHYLOSEQ))[,NAMES])
  sample_data(PHYLOSEQ)$duplic.id<-CATS
  SAMDAT<-sample_data(PHYLOSEQ)
  SAMDAT.sub<-subset(SAMDAT,duplicated(CATS)==F)
  FASTA<-refseq(PHYLOSEQ)
  rownames(SAMDAT.sub)<-SAMDAT.sub$duplic.id
  PHYLOSEQ.merge<-merge_samples(PHYLOSEQ,"duplic.id")
  sample_data(PHYLOSEQ.merge)<-SAMDAT.sub
  PHYLOSEQ.merge<-merge_phyloseq(PHYLOSEQ.merge,FASTA)
  PHYLOSEQ.merge
}

manage_unassigned<-function(PHYLOSEQ,UNASS_STRING=NA,ADD,AFTER=TRUE){
  TAXall<-data.frame(tax_table(PHYLOSEQ),stringsAsFactors = F)
  if(!is.na(UNASS_STRING)){TAXall[UNASS_STRING]<-NA}
  # TAXall<-as.character(TAXall[i,])
  for(i in 1:dim(TAXall)[1]){  
    TAX<-as.character(TAXall[i,])
    if(AFTER==TRUE) {for(j in 2: length(TAX)){
      if(is.na(TAX[j])){TAX[j]<-ifelse(regexpr(ADD,TAX[j-1])>0,
                                       TAX[j-1],paste(TAX[j-1],ADD,sep=""))}}}
    if(AFTER==FALSE) {for(j in 2: length(TAX)){
      if(is.na(TAX[j])){TAX[j]<-ifelse(regexpr(ADD,TAX[j-1])>0,
                                       TAX[j-1],ADD,paste(TAX[j-1],sep=""))}}}
    TAXall[i,]<-TAX
  }
  TAXA<-colnames(TAXall)
  SPECIES<-rownames(TAXall)
  TAXall<-tax_table(TAXall)
  taxa_names(TAXall)<-SPECIES
  colnames(TAXall)<-TAXA
  tax_table(PHYLOSEQ)<-TAXall
  
  #output
  PHYLOSEQ
}

sort(sample_sums(PHYLOSEQ_dupl))
# sample S57_JV81_n_F_trus2R no seq, delete both duplicates
EXCLUDE<-c("S57_JV81_n_F_trus2R","S57_JV81_n_F_trus1R")
PHYLOSEQ_dupl.sub<-prune_samples(!sample_names(PHYLOSEQ_dupl)%in%EXCLUDE,
                                  PHYLOSEQ_dupl)
grepl("S57_JV81_n_F_trus2R",sample_names(PHYLOSEQ_dupl))
grepl("S57_JV81_n_F_trus1R",sample_names(PHYLOSEQ_dupl.sub))

DUPL1<-prune_samples(regexpr("F_trus1R",sample_names(PHYLOSEQ_dupl.sub))>0,
                     PHYLOSEQ_dupl.sub)
DUPL2<-prune_samples(regexpr("F_trus2R",sample_names(PHYLOSEQ_dupl.sub))>0,
                     PHYLOSEQ_dupl.sub)

RICH1<-estimate_richness(DUPL1)
RICH2<-estimate_richness(DUPL2)

head(RICH1)
head(RICH2)

data.frame(sample_names(DUPL1),sample_names(DUPL2))

cor.test(RICH1$Shannon,RICH2$Shannon)
plot(RICH1$Shannon,RICH2$Shannon)

# Prokurstovská analýza duplikátů
DUPL1.prop<-transform_sample_counts(DUPL1,function(x) x/sum(x))
DUPL2.prop<-transform_sample_counts(DUPL2,function(x) x/sum(x))

BC1<-vegdist(otu_table(DUPL1.prop))
BC2<-vegdist(otu_table(DUPL2.prop))

BC1.pc<-pcoa(BC1)$vectors
BC2.pc<-pcoa(BC2)$vectors

PR<-protest(BC1.pc,BC2.pc)
PR

plot(PR)

# splučování duplikatů
CONC<-dupl.concensus(PHYLOS=PHYLOSEQ_dupl.sub,NAMES="ID_individual")
MERGED<-merge.duplicates(PHYLOSEQ = CONC,NAMES="ID_individual")

sum(otu_table(PHYLOSEQ_dupl.sub))
sum(otu_table(CONC))
sum(otu_table(MERGED))

PHYLOSEQ_dupl.sub
MERGED
# save(MERGED,file = "D:/SEQ/Impulsivita/Výsledky/Merged.R")

load("D:/SEQ/Impulsivita/Výsledky/Merged.R")
sort(sample_sums(MERGED))
MERGED<-MERGED
sample_data(MERGED)

# Putting sample 57 back
NEDUPL<-c("S57_JV81_n_F_trus1R")
PHYLOSEQ_dupl.excl<-prune_samples(sample_names(PHYLOSEQ_dupl)%in%NEDUPL,
                                  PHYLOSEQ_dupl)
PHYLOSEQ_dupl.excl.prop<-transform_sample_counts(PHYLOSEQ_dupl.excl,function(x) x/sum(x))
# normalization of sequences - aging of ASVs that are removed during merging of duplicates
# this is done by leaving (FILT1) the sequences that are occupied in those formulas that were sent in duplicates and then (FILT2), the sequences that make up at least 1%
FILT1<-taxa_names(PHYLOSEQ_dupl.excl)%in%taxa_names(MERGED)
FILT2<-apply(otu_table(PHYLOSEQ_dupl.excl.prop),2,max)>0.01

FILT12<-FILT1+FILT2>0
sum(FILT12)

NONDUPL<-prune_taxa(FILT12,PHYLOSEQ_dupl.excl)
PHYLOSEQ.final<-merge_phyloseq(MERGED,NONDUPL)

grep("S57_JV81_n_F_trus1R",sample_names(PHYLOSEQ.final))

# save(PHYLOSEQ.final,file="D:/SEQ/Impulsivita/Výsledky/PHYLOSEQ.final.R")
sort(sample_sums(PHYLOSEQ.final))

#########################################
#Creating OTU table######################
#########################################
View(otu_table(PHYLOSEQ.final))
View(sample_data(PHYLOSEQ.final))
View(tax_table(PHYLOSEQ.final))

library(dplyr)
meta <- sample_data(PHYLOSEQ.final)
taxa <- tax_table(PHYLOSEQ.final)
otu <- otu_table(PHYLOSEQ.final)

PHYLOSEQ.final
dim(otu_table(PHYLOSEQ.final))
DAT<-data.frame(t(otu_table(PHYLOSEQ.final)),tax_table(PHYLOSEQ.final))
DAT$SEQ<-rownames(DAT)
# write.table(DAT,file = "D:/SEQ/Impulsivita/Výsledky/OTU_table_Implusivity.txt",row.names = F,quote = F,sep="\t")
View(DAT)

```

##ALPHA DIVERSITY
```{r cook, echo=FALSE}
load("D:/SEQ/Impulsivita/Výsledky/PHYLOSEQ.final.R")
summary(as.factor(sample_data(PHYLOSEQ.final)$Impulsivity))
PHYLOSEQ.final.rare<-rarefy_even_depth(PHYLOSEQ.final)
PHYLOSEQ.final.prop<-transform_sample_counts(PHYLOSEQ.final,function(x) x/sum(x))
RICH<-estimate_richness(PHYLOSEQ.final.rare)
# write.table(RICH,file="D:/SEQ/Impulsivita/Výsledky/Alpha_diversity.txt", sep="\t",quote = F)
RICH<-data.frame(RICH,sample_data(PHYLOSEQ.final.rare))
# Plots<-ggplot(RICH,aes(x=Impulsivity,y=Shannon))+geom_boxplot(outlier.shape = NA)+geom_jitter() +
  # ggplot(RICH,aes(x=Impulsivity,y=Observed))+geom_boxplot(outlier.shape = NA)+geom_jitter()
# Shannon
plot(RICH$Observed,RICH$Chao1)
model<-lm(Shannon~Impulsivity,data=RICH)
summary(model)
anova(model)
{qqnorm(resid(model))
  qqline(resid(model))}
hist(RICH$Shannon)

model_log<-lm(log10(Shannon)~Impulsivity,data=RICH)
{qqnorm(resid(model_log))
  qqline(resid(model_log))}
log<-(RICH$Shannon)
hist(log)
MASS::boxcox(model)
m0<-lm(Shannon~Impulsivity,data=RICH)
BC<-MASS::boxcox(m0,lambda = seq(-10, 10, 1/10))
lamb<-BC$x[BC$y==max(BC$y)]
if(lamb!=0){TEST.trans<-(RICH$Shannon^lamb-1)/lamb}
if(lamb==0){TEST.trans<-log(RICH$Shannon)}
hist(TEST.trans, main = "Shannon po normalizaci")
QQ<-lm(TEST.trans~Impulsivity,data=RICH)
{qqnorm(resid(QQ))
  qqline(resid(QQ))}
summary(lm(TEST.trans~Impulsivity,data=RICH))
summary(aov(TEST.trans~Impulsivity,data=RICH))
summary(lm(log10(Shannon)~Impulsivity,data=RICH))
summary(aov(log10(Shannon)~Impulsivity,data=RICH))
kruskal.test(Shannon~Impulsivity,data=RICH)

# Observed
model<-lm(Observed~Impulsivity,data=RICH)
summary(model)
anova(model)
{qqnorm(resid(model))
  qqline(resid(model))}
hist(RICH$Observed)
model_log<-lm(log10(Observed)~Impulsivity,data=RICH)
{qqnorm(resid(model_log))
  qqline(resid(model_log))}
log<-log10(RICH$Observed)
hist(log)
MASS::boxcox(model)
m0<-lm(Observed~Impulsivity,data=RICH)
BC<-MASS::boxcox(m0,lambda = seq(-10, 10, 1/10))
lamb<-BC$x[BC$y==max(BC$y)]
if(lamb!=0){TEST.trans<-(RICH$Observed^lamb-1)/lamb}
if(lamb==0){TEST.trans<-log(RICH$Observed)}
hist(TEST.trans, main = "Observed po normalizaci")
QQ<-lm(TEST.trans~Impulsivity,data=RICH)
{qqnorm(resid(QQ))
  qqline(resid(QQ))}
summary(lm(TEST.trans~Impulsivity,data=RICH))
summary(aov(TEST.trans~Impulsivity,data=RICH))
summary(lm(log10(Observed)~Impulsivity,data=RICH))
summary(aov(log10(Observed)~Impulsivity,data=RICH))
kruskal.test(Observed~Impulsivity,data=RICH)

# ALPHA DIVERZITY - GRAFY
library(ggplot2)
library(gridExtra)
setwd("D:/SEQ/Impulsivita/Výsledky")
plot1 <- ggplot(RICH, aes(x = Impulsivity, y = Shannon, color=Impulsivity,fill=Impulsivity)) +
  geom_violin(alpha=0.3,draw_quantiles=0.5) +
  geom_jitter() +
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 12),
        legend.position = "none", legend.title = element_blank()) +
  xlab(NULL)

plot2 <- ggplot(RICH, aes(x = Impulsivity, y = Observed,color=Impulsivity,fill=Impulsivity)) +
  geom_violin(alpha=0.3,draw_quantiles=0.5) +
  geom_jitter() +
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 12))+
  xlab(NULL)+
  guides(fill = guide_legend(title = "Groups"))

combined_plot <- grid.arrange(plot1, plot2, ncol = 2)
print(combined_plot)

# ggsave("Alpha diversity_fill+color.png", plot = combined_plot, dpi = 300)

```

```{r wildflower, echo=FALSE}
BC<-vegdist(otu_table(PHYLOSEQ.final.prop))
JA<-vegdist(data.frame(otu_table(PHYLOSEQ.final.rare)),method = "jaccard",binary = T)

ord<-ordinate(PHYLOSEQ.final.prop,method = "PCoA", BC)
plot_ordination(PHYLOSEQ.final.prop,ord,color="Impulsivity")
plot_ordination(PHYLOSEQ.final.prop,ord,color="Impulsivity")+facet_wrap(.~Impulsivity)

ord<-ordinate(PHYLOSEQ.final.prop,method = "PCoA", JA)
plot_ordination(PHYLOSEQ.final.prop,ord,color="Impulsivity")
plot_ordination(PHYLOSEQ.final.prop,ord,color="Impulsivity")+facet_wrap(.~Impulsivity)

SD<-data.frame(sample_data(PHYLOSEQ.final.prop))
adonis2(BC~Impulsivity,data=SD)
adonis2(JA~Impulsivity,data=SD)

# Betadisper
BETA<-betadisper(JA,SD$Impulsivity)
anova(BETA)
boxplot(BETA)
TukeyHSD(BETA)

# BETA DIVERZITA - GRAFY
install.packages("ggpubr")
library(ggpubr)

small.fig<-function(x){x+theme_bw()+
    theme(axis.text = element_text(size=8),
          axis.title = element_text(size=8),
          strip.text = element_text(size=8),
          legend.text = element_text(size=8),
          legend.title = element_text(size=8))}

prepare_spiders<-function(ORD,PHYLO,AXES,CAT){
  VECT<-data.frame(ORD$vectors,sample_data(PHYLO))
  LIST<-list()
  for(i in 1:length(CAT)){
    LIST[[i]]<-VECT[,CAT[i]]
  }
  
  CENTROIDx<-tapply(VECT[,AXES[1]],LIST,mean)
  CENTROIDy<-tapply(VECT[,AXES[2]],LIST,mean)
  
  CENTROIDx.l<-reshape2::melt(CENTROIDx)
  CENTROIDy.l<-reshape2::melt(CENTROIDy)
  
  names(CENTROIDx.l)<-c(CAT,"centr.x")
  names(CENTROIDy.l)<-c(CAT,"centr.y")
  
  CENTROIDxy.l<-data.frame(CENTROIDx.l,centr.y=CENTROIDy.l[,dim(CENTROIDy.l)[2]])
  VECT.l<-join(VECT,CENTROIDxy.l)
  VECT.l
}

Hull_get<-function(ORD,VAR){
  SCORES<-ORD$data
  SCORES$VAR<-SCORES[,VAR]
  SCORES$Scores1<-SCORES[,1]
  SCORES$Scores2<-SCORES[,2]
  
  hull_cyl <- SCORES %>%
    group_by(VAR) %>%
    dplyr::slice(chull(Scores1, Scores2))
  hull_cyl
}


ord.bc<-ordinate(PHYLOSEQ.final.rare,method = "PCoA", BC)
ord.bc.p<-plot_ordination(PHYLOSEQ.final.rare,ord.bc,color="Impulsivity")

ord.ja<-ordinate(PHYLOSEQ.final.rare,method = "PCoA", JA)
ord.ja.p<-plot_ordination(PHYLOSEQ.final.rare,ord.ja,color="Impulsivity")

HULL<-Hull_get(ORD=ord.bc.p, VAR="Impulsivity")
ord.bc.p<-ord.bc.p+ geom_polygon(data = HULL,aes(fill=Impulsivity),alpha=0.1)

HULL<-Hull_get(ORD=ord.ja.p, VAR="Impulsivity")
ord.ja.p<-ord.ja.p+ geom_polygon(data = HULL,aes(fill=Impulsivity),alpha=0.1)

Spider_df.bc<-prepare_spiders(ORD=ord.bc, PHYLO=PHYLOSEQ.final.rare,AXES=c(1,2),CAT=c("Impulsivity"))
Spider_df.ja<-prepare_spiders(ORD=ord.ja, PHYLO=PHYLOSEQ.final.rare,AXES=c(1,2),CAT=c("Impulsivity"))

ord.bc.p<-ord.bc.p+geom_segment(data=Spider_df.bc,aes(x = centr.x, y = centr.y, xend = Axis.1, yend =Axis.2),
                                arrow = arrow(length = unit(0.0, "cm")),alpha=0.2)

ord.ja.p<-ord.ja.p+geom_segment(data=Spider_df.ja,aes(x = centr.x, y = centr.y, xend = Axis.1, yend =Axis.2),
                                arrow = arrow(length = unit(0.0, "cm")),alpha=0.2)

ord.bc.s<-small.fig(ord.bc.p)
ord.ja.s<-small.fig(ord.ja.p)

ord.bc.s<-ord.bc.s+theme(legend.title = element_blank())
ord.ja.s<-ord.ja.s+theme(legend.title = element_blank())

GRID.pc<-ggarrange(ord.bc.s+ggtitle("A) Bray - Curtis")+theme(plot.title = element_text(size = 12,face = "bold")),
                   ord.ja.s+ggtitle("B) Jaccard")+theme(plot.title = element_text(size = 12,face = "bold")),
                   nrow = 2,
                   common.legend = T,legend = "bottom")
GRID.pc
# ggsave("Beta diverzita.png", plot = GRID.pc, dpi = 300, width = 8, height = 8)

GRID.pc

```

##TAXONOMY GRAPHS
```{r cangaroo, echo=FALSE}
# Tax profiles
prepare_tax_df<-function(PHYLOSEQ,RANK,Unass.symbol=NA,Unass.repl,
                         min_prop,top_tax,bellow_top,
                         merge_categories,prop.trans,sort_abu){
  if(prop.trans==T){PHYLOSEQ=transform_sample_counts(PHYLOSEQ,function(x) x/sum(x))}
  TT<-tax_table(PHYLOSEQ)
  class(TT)<-"matrix"
  TT[is.na(tax_table(TT))]<-Unass.repl
  TT<-tax_table(TT)
  tax_table(PHYLOSEQ)<-TT
  
  PHYLOSEQ.merged<-tax_glom(PHYLOSEQ,taxrank=RANK)
  PHYLOSEQ.merged=transform_sample_counts(PHYLOSEQ.merged,function(x) x/sum(x))
  
  #select top taxa
  NAMES<-as.character(tax_table(PHYLOSEQ.merged)[,RANK])
  TS<-taxa_sums(PHYLOSEQ.merged)
  names(TS)<-NAMES
  TS<-rev(sort(TS))
  KEEP<-names(TS)[1:top_tax]
  
  #
  TT<-tax_table(PHYLOSEQ.merged)
  class(TT)<-"matrix"
  TT[,RANK][!TT[,RANK]%in%KEEP]<-bellow_top
  TT<-tax_table(TT)
  tax_table(PHYLOSEQ.merged)<-TT
  PHYLOSEQ.merged<-tax_glom(PHYLOSEQ.merged,taxrank=RANK)
  
  mdf = psmelt(PHYLOSEQ.merged)
  
  if(sort_abu==T){
    mdf[,RANK]<-as.factor(mdf[,RANK])
    ADD<-levels(mdf[,RANK])[!levels(mdf[,RANK])%in%KEEP]
    if(length(ADD)>0){mdf[,RANK]<-factor(mdf[,RANK],levels=c(KEEP,ADD))}
    if(length(ADD)==0){mdf[,RANK]<-factor(mdf[,RANK],levels=c(KEEP))}
  }
  
  mdf
}

manage_unassigned<-function(PHYLOSEQ,UNASS_STRING=NA,ADD,AFTER=TRUE){
  TAXall<-data.frame(tax_table(PHYLOSEQ),stringsAsFactors = F)
  if(!is.na(UNASS_STRING)){TAXall[UNASS_STRING]<-NA}
  # TAXall<-as.character(TAXall[i,])
  for(i in 1:dim(TAXall)[1]){  
    TAX<-as.character(TAXall[i,])
    if(AFTER==TRUE) {for(j in 2: length(TAX)){
      if(is.na(TAX[j])){TAX[j]<-ifelse(regexpr(ADD,TAX[j-1])>0,
                                       TAX[j-1],paste(TAX[j-1],ADD,sep=""))}}}
    if(AFTER==FALSE) {for(j in 2: length(TAX)){
      if(is.na(TAX[j])){TAX[j]<-ifelse(regexpr(ADD,TAX[j-1])>0,
                                       TAX[j-1],ADD,paste(TAX[j-1],sep=""))}}}
    TAXall[i,]<-TAX
  }
  TAXA<-colnames(TAXall)
  SPECIES<-rownames(TAXall)
  TAXall<-tax_table(TAXall)
  taxa_names(TAXall)<-SPECIES
  colnames(TAXall)<-TAXA
  tax_table(PHYLOSEQ)<-TAXall
  
  #output
  PHYLOSEQ
}

phyloseq_2_GG<-function(which.taxa,which.phyloseq,manage.ussigned=NULL,unassign.string=NULL){  
  RESULT<-data.frame(stringsAsFactors = F)
  SSUMS<-sample_sums(which.phyloseq)
  for(i in 1: length(which.taxa))
  {
    actual.phylos<-prune_taxa(taxa_names(which.phyloseq)%in%which.taxa[i],which.phyloseq)
    Abundance<-as.numeric(otu_table(actual.phylos))  
    OTU<-rep(which.taxa[i],length(Abundance))
    Seq.tot<-rep(SSUMS[i],length(Abundance))
    Taxo.actual<-as.character(tax_table(actual.phylos))
    if(manage.ussigned==T){
      for(j in 2: length(Taxo.actual)){
        if(is.na(Taxo.actual[j])){Taxo.actual[j]<-ifelse(regexpr(unassign.string,Taxo.actual[j-1])>0,
                                                         Taxo.actual[j-1],paste(Taxo.actual[j-1],unassign.string,sep=""))}
      }
    }
    
    Taxo.actual.rbind<-do.call("rbind", replicate(length(sample_sums(actual.phylos)), Taxo.actual, simplify = FALSE))
    colnames(Taxo.actual.rbind)<-colnames(tax_table(which.phyloseq))
    Sd<-sample_data(actual.phylos)
    TEMP<-data.frame(Abundance,OTU,Seq.tot,Taxo.actual.rbind,Sd,stringsAsFactors = F)
    RESULT<-rbind(RESULT,TEMP)
  }
  RESULT
}

PHYLOSEQ.final.prop_tax<-manage_unassigned(PHYLOSEQ=PHYLOSEQ.final.prop,
                                                UNASS_STRING=NA,ADD="",AFTER=TRUE)

DF.tax.caecum.class<-prepare_tax_df(PHYLOSEQ=PHYLOSEQ.final.prop_tax,
                                    RANK="Class",Unass.symbol=NA,
                                    Unass.repl="Unassigned",min_prop=NULL,top_tax=9,
                                    bellow_top="Others",merge_categories=NULL,
                                    prop.trans=TRUE,sort_abu=TRUE)
library(Polychrome)
P25<-createPalette(25, c("#0dff00", "#ff0000"), M=1000) #palette for taxplots
names(P25)<-NULL

#Class
ta3 = ggplot(DF.tax.caecum.class, aes_string(x = "ID_individual", y = "Abundance", fill = "Class",order="Class"))+theme_bw(base_size = 12)
ta3 = ta3 + geom_bar(stat = "identity", position = "stack")
ta3= ta3 + theme(axis.title.x=element_blank(),axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(hjust = 0,size=10))
ta3 <- ta3 + facet_grid(.~Impulsivity, scales = "free",space="free")+
  theme(strip.text = element_text(size = 10, angle = 0))+
  scale_fill_manual(values = P25)
gg.class<-ta3

#Genus
DF.tax.caecum.genus<-prepare_tax_df(PHYLOSEQ=PHYLOSEQ.final.prop_tax,
                                    RANK="Genus",Unass.symbol=NA,
                                    Unass.repl="Unassigned",min_prop=NULL,top_tax=23,
                                    bellow_top="Others",merge_categories=NULL,
                                    prop.trans=T,sort_abu=TRUE)

ta3 = ggplot(DF.tax.caecum.genus, aes_string(x = "ID_individual", y = "Abundance", fill = "Genus",order="Genus"))+theme_bw(base_size = 12)
ta3 = ta3 + geom_bar(stat = "identity", position = "stack")
ta3= ta3 + theme(axis.title.x=element_blank(),axis.text.x = element_text(size = 7, angle = 90),axis.text.y = element_text(hjust = 0,size=10))
ta3 <- ta3 + facet_grid(.~Impulsivity, scales = "free",space="free")+
  theme(strip.text = element_text(size = 10, angle = 0))+
  scale_fill_manual(values = P25)
gg.genus<-ta3

gg.class.s<-small.fig(gg.class)
gg.genus.s<-small.fig(gg.genus)

gg.class.s<-gg.class.s+theme(axis.text.x = element_blank(),
                             axis.title.x = element_blank())


gg.genus.s<-gg.genus.s+theme(axis.text.x = element_blank(),
                             axis.title.x = element_blank())

# ggsave("Taxonomy_class_bez ID.png", plot = gg.class.s, dpi = 300)
# ggsave("Taxonomy_genus_bez ID.png", plot = gg.genus.s, dpi = 300)


```

##DESEQ
```{r FISH, echo= FALSE}
library(BiocManager)
BiocManager::install("DESeq2")
library(DESeq2)

diagdds = phyloseq_to_deseq2(PHYLOSEQ.final, ~ Impulsivity)

gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

geoMeans = apply(counts(diagdds), 1, gm_mean)
diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
diagdds = DESeq(diagdds)

res=results(diagdds,cooksCutoff=T)

res<-data.frame(res,tax_table(PHYLOSEQ.final))

alpha = 0.05
sigtab = res[which(res$padj < alpha), ]
sigtab

DD<-phyloseq_2_GG(which.taxa=rownames(sigtab),
                  which.phyloseq=PHYLOSEQ.final,
                  manage.ussigned=T,
                  unassign.string="_unass.")
DD$NAMES<-paste(DD$OTU,DD$Genus,sep="\n")
DD$genus<-DD$Genus
DD$genus<-gsub("_unass.","",DD$genus)
DD$genus<-gsub(".unclassif","",DD$genus)

GG.diff<-ggplot(DD,aes(y=Abundance/Seq.tot,x=Impulsivity))+geom_boxplot(outlier.color = NA)+geom_jitter(alpha=0.3)+facet_wrap(.~Genus)
GG.diff.s<-small.fig(GG.diff)
GG.diff.s<-GG.diff.s+theme(axis.title.x = element_blank())+ylab("Relative abundance")
GG.diff.s

# ggsave("DESEQ_IMP.png", plot = GG.diff.s, dpi = 300, width = 8, height = 8)

```

